### ISSUE TITLE
Forbidden - referer invalid

### ISSUE THREAD

**Issue by NextdoorSec** [gophish/gophish#2765]
Thanks for reaching out! We're happy to help resolve issues as quickly as possible.

Please use this template when creating a new issue. If you do not follow this template format, your issue may be closed without comment.

Before filing a new issue, please use the search bar at the top of the browser to search for similar issues. Also, please make sure you have read our [documentation](https://getgophish.com/documentation/), which covers how to use Gophish.

What version of Gophish are you using?: v0.12.1 (built from source), tried other ways as well

Brief description of the issue: Forbidden - referer invalid

What are you expecting to see happen? : dashboard page after entering my credentials

What are you seeing happen? Forbidden - referer invalid

If this question is related to email templates or landing pages not working as expected, please provide your template or landing page below:

Insert code here
Please provide any terminal output that may be relevant below:

time="2023-02-15T15:45:06Z" level=info msg="2a01:5241:665:1300::84f0 - - [15/Feb/2023:15:45:06 +0000] \"POST /login?next=%2F HTTP/1.1\" 403 28 \"https://<domain>/login?next=%2F\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36\""

Insert output here
Please provide as many steps as you can to reproduce the problem:

Step 1 - sudo ./gophish
Step 2 - error while entering credentials from reverse proxy, which is cloudflared:
https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/install-and-setup/tunnel-guide/local/ 

It works fine from the IP of my VPS directly, I've tried adding csrf_keys to the config.json file, but no difference. Tried setting noTLSVerify key to "false", but no difference. I've tried adding the subdomain, main domain, all the possible origins in the trusted_origins key, but still no difference. Have probably read anything here regarding that issue, but still no luck;

**Comment by glennzw** [comment]
Hi there. If you're behind a reverse proxy you'll want to use the trusted_origins parameter:

https://docs.getgophish.com/user-guide/installation

**Comment by NextdoorSec** [comment]
Hi Glenn,

I've tried using the "trusted_origins" param as well, but it didn't make a difference. I've even tried giving in cloudflare's nameservers as well. Then restarting gophish as usual and trying from an incognito tab.

**Comment by glennzw** [comment]
Could you share your config.json and terminal output and I'll see if I can spot anything anomalous.

**Comment by NextdoorSec** [comment]
{
        "admin_server": {
                "listen_url": "0.0.0.0:1995",
                "use_tls": true,
                "cert_path": "gophish_admin.crt",
                "key_path": "gophish_admin.key",
                "trusted_origins": ["TUNNEL_ID.cfargotunnel.com", "https://VPS_IP/, "main_domain.com", "athena.ns.cloudflare.com", "vin.ns.cloudflare.com"]
        },
        "phish_server": {
                "listen_url": "0.0.0.0:80",
                "use_tls": false,
                "cert_path": "example.crt",
                "key_path": "example.key"
        },
        "db_name": "sqlite3",
        "db_path": "gophish.db",
        "migrations_prefix": "db/db_",
        "contact_address": "",
        "logging": {
                "filename": "",
                "level": ""
        }
}

**Comment by glennzw** [comment]
Sorry @ndsecc this one slipped through. I've only tested trusted_origins using ngrok, could you try it with that so we establish a base working case, and then figure out what's different with Cloudflare?

**Comment by avicoder** [comment]
any lessons to learn? I am still facing the issue when using cloudflare tunnel :/

**Comment by remyxed** [comment]
I'm having the same issue when using a cloudflare tunnel. 
I tried everything as trusted_domains, including using wildcards, but to no avial. 

The weird thing is, when I go to the domain address, I do see the GoPHish login screen, but it is after logging in I get the "Forbidden - referer invalid" error.

**Comment by yannispgs** [comment]
Did you try with the hostname/domain only ? `domainname` without the port, nor `/*` ? When I tried all the syntaxes, adding any prefix (protocol) or suffix (path, wildcard, port number, etc.) kept me from logging in correctly.

**Comment by timoa** [comment]
Hello Guys!
After hitting a wall with cloudflared and the CSRF referer issue you're facing too, I found a solution!

I saw other people had issues with Cloudflared and Django or other Web apps supporting CSRF.
One solves the issue using Nginx between the Cloudflare Tunnel and the Web app.

Since I wanted to have multiple domains and Let's Encrypt support for the landing pages, I installed it and tried to rewrite the Nginx proxy params to allow cloudflared requests and it works! 🎉

I hope that will help!

**GoPhish Admin UI**
 Cloudflare Tunnel <=> Nginx (8081) <=> GoPhish Admin UI (3333)

**GoPhish Public**
Nginx (80/443) <=> GoPhish Public (8080)

**GoPhish config**
```json
{
  "admin_server": {
    "listen_url": "0.0.0.0:3333",
     "use_tls": false,
     "cert_path": "gophish_admin.crt",
     "key_path": "gophish_admin.key",
     "trusted_origins": [
       "http://localhost",
       "https://gophish.example.com"
     ]
   },
   "phish_server": {
      "listen_url": "127.0.0.1:8080",
      "use_tls": false,
      "cert_path": "example.crt",
      "key_path": "example.key"
    },
    "db_name": "mysql",
    "db_path": "USER:PASS@(ENDPOINT:3306)/DB?charset=utf8&parseTime=True&loc=UTC",
    "migrations_prefix": "db/db_",
    "contact_address": "email@example.com",
    "logging": {
      "filename": "/var/log/gophish.log",
       "level": "info"
     }
  }
  ```

**Cloudflared config**
```yaml
tunnel: TUNNEL_ID
credentials-file: /etc/cloudflared/credentials.json
logfile: /var/log/cloudflared.log
loglevel: info

ingress:

  # GoPhish Admin UI
  - hostname: gophish.example.com
    service: http://localhost:8081

  # GoFish MySQL DB
  - hostname: gophish-db.example.com
    service: tcp://ENDPOINT:3306

  # Default service
  - service: http_status:404
 ```

**Nginx GoPhish Public**
```
server {
  listen 80 default_server;
  listen [::]:80 default_server;

  server_name example.com test.example.com;

  location / {
    proxy_pass http://127.0.0.1:8080;
    include proxy_params;
  }
}
```

**Nginx GoPhish Admin UI**
```
server {
  listen 8081;
  listen [::]:8081;

  server_name localhost;

  location / {
    proxy_pass http://localhost:3333;
    proxy_ssl_verify off;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

    # Set for CSRF bug that results in "403 Forbidden - referer invalid"
    proxy_set_header Referer "";
    proxy_set_header X-Forwarded-Proto "";

    #include proxy_params;
  }
}
```

**Comment by sam-skinner-notarize** [comment]
@timoa Any reason this wouldn't work with an ALB in AWS? I have the config (json) the same as yours but I'm still getting the referer invalid error.

**Comment by home4lab** [comment]
in my case, i just set trusted_origins with subdomain that i want to use, then my traefik doesnt show error Forbidden - referer invalid
```{
        "admin_server": {
                "listen_url": "0.0.0.0:3333",
                "use_tls": true,
                "cert_path": "gophish_admin.crt",
                "key_path": "gophish_admin.key",
                "trusted_origins": ["yourdomain.com"]
        },
        "phish_server": {
                "listen_url": "0.0.0.0:80",
                "use_tls": false,
                "cert_path": "example.crt",
                "key_path": "example.key"
        },
        "db_name": "sqlite3",
        "db_path": "gophish.db",
        "migrations_prefix": "db/db_",
        "contact_address": "",
        "logging": {
                "filename": "",
                "level": ""
        }
}
```

**Comment by 5pr1nz** [comment]
`Forbidden - referer invalid` usually occurs due to `X-Forwarded-Proto https` set in the request headers. After disabling this header, gophish works like a charm.

A forward proxy might be able to unset this header from upstream - if required.

**Comment by xtphantom** [comment]
tl;dr; "trusted_origins": ["mysite.example.com"]
The csrf.go code is looking for host name only with subdomain if used.  setting trusted_domains to "mysite.example.com", no https or other url parts.
Actual example from csrf.go 
if referer.Host == trustedOrigin {
	valid = true
	break
}

**Comment by 5pr1nz** [comment]
Understood @xtphantom and all correct. However, [the initial question](https://github.com/gophish/gophish/issues/2765#issue-1586097222) about the error `Forbidden - referer invalid` occurred to me - also while setting the trusted_origins - when `X-Forwarded-Proto https` was set. This header is usually set by upstream proxies and it is unclear to me why Gophish responds in this way.

**Comment by deduced** [comment]
Ok I had this exact issue when trying to use gophish with ngrok. Not too much info on this and many recommend using trusted_origins. IN my case I had to remove the X-Forwarded-Proto header: 

`ngrok http https://localhost:3333 --request-header-remove="X-Forwarded-Proto"`

Hope this helps someone.

**Comment by ljrk0** [comment]
When `X-Forwarded-Proto` is set by the reverse proxy, gophish enforces that `X-Forwarded-Host` matches an entry in the `trusted_origins` array. This would usually be in the format of domain:port. To verify this, simply print a request by the reverse proxy and review the header. E.g., for the Caddy reverse proxy I used a `Caddyfile` such as:

```
foobar.example.com:8443 {
        reverse_proxy :4444
}
```

and was listening on my server using netcat on port 4444. You'll get some output like:

```
$ nc -l 4444
GET / HTTP/1.1
Host: foobar.example.com:8443
[...]
Upgrade-Insecure-Requests: 1
Via: 2.0 Caddy
X-Forwarded-For: cli.ent.ip.adr
X-Forwarded-Host: foobar.example.com:8443
X-Forwarded-Proto: https
```

Adding `foobar.example.com:8443` fixed the invalid referer. If you remove the `X-Forwarded-Proto` header using the proxy, the check won't run and the invalid referer will also go away, naturally.

**Comment by npulidom** [comment]
this is the way to go with AWS ALB
```
 "admin_server": {
                "listen_url": "0.0.0.0:3333",
                "use_tls": true,
                "cert_path": "gophish_admin.crt",
                "key_path": "gophish_admin.key",
                "trusted_origins": ["yourdomain.com"]
        }
```

**Comment by imamuppet** [comment]
Wanted to thank npulidom, and everyone else.  Can confirm the "yourdomain.com" in trusted_origins works behind haproxy.
